src/pages/Compare/index.tsx

```diff
@@ -62,43 +70,66 @@ export class CompareBase extends React.Component<Props> {
   }
 
   loadData(prevProps?: Props) {
-    const { match } = this.props;
+    const { match, path } = this.props;
     const { addonId, baseVersionId, headVersionId } = match.params;
 
     if (
       !prevProps ||
+      path !== prevProps.path ||
```

Comment:
In your current patch at http://localhost:3000/en-US/compare/502955/versions/1541786...1541798/, the network request to fetch the diff happens twice which isn't great. The first fetch happens without a path and the second one happens with a path. 

I think, also, the fetch should not be happening while one is in progress. Below is a patch that only does the fetch whe `isLoading` is `false`. This stops the double requesting from happening but it might not be the right thing to do. In any case, it would mean that `null` can't be used to signify an error state.

```diff
diff --git a/src/pages/Compare/index.tsx b/src/pages/Compare/index.tsx
index 3b8ae14..704e092 100644
--- a/src/pages/Compare/index.tsx
+++ b/src/pages/Compare/index.tsx
@@ -70,15 +70,16 @@ export class CompareBase extends React.Component<Props> {
   }
 
   loadData(prevProps?: Props) {
-    const { match, path } = this.props;
+    const { isLoading, match, path } = this.props;
     const { addonId, baseVersionId, headVersionId } = match.params;
 
     if (
-      !prevProps ||
-      path !== prevProps.path ||
-      addonId !== prevProps.match.params.addonId ||
-      baseVersionId !== prevProps.match.params.baseVersionId ||
-      headVersionId !== prevProps.match.params.headVersionId
+      !isLoading &&
+      (!prevProps ||
+        path !== prevProps.path ||
+        addonId !== prevProps.match.params.addonId ||
+        baseVersionId !== prevProps.match.params.baseVersionId ||
+        headVersionId !== prevProps.match.params.headVersionId)
     ) {
       const { dispatch, _fetchDiff } = this.props;
 
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 6d3ea48..adeea9e 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -233,8 +233,8 @@ export type VersionsState = {
   };
   compareInfo:
     | CompareInfo // data successfully loaded
-    | null // an error has occured
-    | undefined; // data not fetched yet
+    | null // data fetching has not begun or an error occurred
+    | undefined; // data fetching is in progress
   versionInfo: {
     [versionId: number]: Version;
   };
@@ -247,7 +247,7 @@ export type VersionsState = {
 
 export const initialState: VersionsState = {
   byAddonId: {},
-  compareInfo: undefined,
+  compareInfo: null,
   versionFiles: {},
   versionInfo: {},
 };
```

Reply:
@kumar303 what about this:

```diff
diff --git a/src/pages/Compare/index.tsx b/src/pages/Compare/index.tsx
index 3b8ae14..0e33d7a 100644
--- a/src/pages/Compare/index.tsx
+++ b/src/pages/Compare/index.tsx
@@ -75,7 +75,7 @@ export class CompareBase extends React.Component<Props> {
 
     if (
       !prevProps ||
-      path !== prevProps.path ||
+      (prevProps.path && path !== prevProps.path) ||
       addonId !== prevProps.match.params.addonId ||
       baseVersionId !== prevProps.match.params.baseVersionId ||
       headVersionId !== prevProps.match.params.headVersionId
```

This fixes the issue on the first load.

Reply:
I pushed a new commit with a test case.